catalog/app/utils/Dialogs.tsx

```diff
@@ -0,0 +1,67 @@
+import * as React from 'react'
+import * as M from '@material-ui/core'
+
+import defer from 'utils/defer'
+import type { Resolver } from 'utils/defer'
+
+type DialogState = 'open' | 'closing' | 'closed'
+
+type ExtraDialogProps = Omit<M.DialogProps, 'open' | 'onClose' | 'onExited'>
+
+export type Close<R> = [R] extends [never] ? () => void : Resolver<R>['resolve']
+
+type Render<R> = (props: { close: Close<R> }) => JSX.Element
+
+interface Dialog {
+  render: Render<any>
+  props?: ExtraDialogProps
+  resolver: Resolver<any>
+}
+
+export const useDialogs = () => {
+  const [state, setState] = React.useState<DialogState>('closed')
+  const [dialog, setDialog] = React.useState<Dialog | null>(null)
+
+  const open = React.useCallback(
+    <R = never,>(render: Render<R>, props?: ExtraDialogProps) => {
+      const { resolver, promise } = defer<R>()
+      setDialog({ render, props, resolver })
+      setState('open')
+      return promise
+    },
+    [setDialog, setState],
+  )
+
+  const close = React.useCallback(
+    (reason: any) => {
```

Comment:
Aren't there just three "reasons" 'cancel' | 'empty' and 'readme'?

Reply:
So, maybe, you can use `M.DialogReason` (I don't remember type exactly, but there is something like this), and then assert for "reasons" we support.

Reply:
the idea is to make the return value configurable

> 'cancel' | 'empty' and 'readme'?

yeah, exactly, these are the "reasons" for this exact dialog -- they may be different for the other cases (in most cases we don't need that at all -- defaulting to "never")